File: src/main/java/dev/xdbl/xdblarenas/players/PlayerManager.java
package dev.xdbl.xdblarenas.players;

import dev.xdbl.xdblarenas.XDBLArena;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.configuration.file.YamlConfiguration;

import java.io.File;
import java.io.IOException;
import java.util.*;

public class PlayerManager {

    private final XDBLArena plugin;
    private final List<ArenaPlayer> arenaPlayers = new ArrayList<>();

    public PlayerManager(XDBLArena plugin) {
        this.plugin = plugin;

        // Load arenaPlayers
        File f = new File(plugin.getDataFolder(), "data/players");
        if (!f.exists()) {
            f.mkdirs();
            return;
        }

        File[] files = f.listFiles();
        for (File file : files) {
            FileConfiguration config = YamlConfiguration.loadConfiguration(file);

            ArenaPlayer arenaPlayer = new ArenaPlayer(
                    plugin,
                    UUID.fromString(file.getName().split("\\.")[0]),
                    config.getInt("elo", 1500),
                    config.getBoolean("scoreboard", false),
                    config.getBoolean("arenabanned", false),
                    config.getBoolean("pvpbanned", false),
                    config.getInt("wins", 0),
                    config.getInt("losses", 0),
                    config.getInt("draws", 0),
                    config.getStringList("logs"),
                    config.getInt("lastSeen", (int) (System.currentTimeMillis() / 1000L)),
                    file
            );
            arenaPlayers.add(arenaPlayer);
        }
    }

    public List<ArenaPlayer> getArenaPlayers() {
        return arenaPlayers;
    }

    public ArenaPlayer getArenaPlayer(UUID playerUUID) {
        // Check if the player is already in the list
        for (ArenaPlayer arenaPlayer : arenaPlayers) {
            if (arenaPlayer.getUUID().equals(playerUUID)) {
                return arenaPlayer;
            }
        }

        // If the player is not in the list, create a new ArenaPlayer
        ArenaPlayer newPlayer = createNewArenaPlayer(playerUUID);
        arenaPlayers.add(newPlayer);

        return newPlayer;
    }

    public void PlayerKill(UUID playerKiller, UUID playerVictim) {
        ArenaPlayer killer = getArenaPlayer(playerKiller);
        ArenaPlayer victim = getArenaPlayer(playerVictim);

        // Constants for the Elo calculation
        double kFactor = 32.0;
        double expectedScoreKiller = 1.0 / (1.0 + Math.pow(10, (victim.getElo() - killer.getElo()) / 400.0));
        double expectedScoreVictim = 1.0 / (1.0 + Math.pow(10, (killer.getElo() - victim.getElo()) / 400.0));

        // Update Elo ratings
        int newEloKiller = (int) (killer.getElo() + kFactor * (1.0 - expectedScoreKiller));
        int newEloVictim = (int) (victim.getElo() + kFactor * (0.0 - expectedScoreVictim));

        // Set new Elo ratings
        killer.setElo(newEloKiller);
        victim.setElo(newEloVictim);
    }

    public int CalculateWinChance(UUID playerKiller, UUID playerVictim){
        ArenaPlayer killer = getArenaPlayer(playerKiller);
        ArenaPlayer victim = getArenaPlayer(playerVictim);

        double expectedScoreKiller = 1.0 / (1.0 + Math.pow(10, (victim.getElo() - killer.getElo()) / 400.0));
        int winChance = (int) (expectedScoreKiller * 100);

        return winChance;
    }

    private ArenaPlayer createNewArenaPlayer(UUID playerUUID) {
        File configFile = new File(plugin.getDataFolder(), "data/players/" + playerUUID + ".yml");
        try {
            configFile.createNewFile();
            YamlConfiguration config = YamlConfiguration.loadConfiguration(configFile);

            // Set default values or load from other sources as needed
            int defaultElo = 1500;

            config.set("elo", defaultElo);
            config.set("arenabanned", false);
            config.set("pvpbanned", false);
            config.set("scoreboard", false);
            config.set("wins", 0);
            config.set("losses", 0);
            config.set("logs", new ArrayList<String>());

            config.save(configFile);

            return new ArenaPlayer(
                    plugin,
                    playerUUID,
                    defaultElo,
                    false,
                    false,
                    false,
                    0,
                    0,
                    0,
                    new ArrayList<String>(),
                    (int) (System.currentTimeMillis() / 1000L),
                    configFile);
        } catch (IOException e) {
            e.printStackTrace();
            return null; // Handle the exception based on your needs
        }
    }
}

File: src/main/java/com/ceraia/managers/PlayerManager.kt
package com.ceraia.managers

import com.ceraia.Ceraia
import com.ceraia.types.CeraiaPlayer
import net.kyori.adventure.text.minimessage.MiniMessage
import org.bukkit.Bukkit
import org.bukkit.configuration.file.FileConfiguration
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.entity.Player
import org.bukkit.scoreboard.DisplaySlot
import org.bukkit.scoreboard.Objective
import org.bukkit.scoreboard.Scoreboard
import org.bukkit.scoreboard.ScoreboardManager
import java.io.File
import java.io.IOException
import java.util.*

class PlayerManager(private val plugin: Ceraia) {
    val ceraiaPlayers: MutableList<CeraiaPlayer> = mutableListOf()

    init {
        // Load arenaPlayers
        val dataFolder = File(plugin.dataFolder, "data/users")
        if (!dataFolder.exists()) {
            dataFolder.mkdirs()
        }

        dataFolder.listFiles()?.forEach { file ->
                val config: FileConfiguration = YamlConfiguration.loadConfiguration(file)

            if(config.getString("uuid") == null) {

            }

            val ceraiaPlayer = CeraiaPlayer(
                    plugin,
                    config.getString("name") ?: throw IllegalArgumentException("Name cannot be null"),
                    config.getString("race", "human").toString(),
                    config.getString("faction"),
                    config.getString("married"),
                    UUID.fromString(config.getString("uuid") ?: throw IllegalArgumentException("UUID cannot be null")),
                    config.getInt("elo", 1500),
                    config.getBoolean("arenabanned", false),
                    config.getBoolean("pvpbanned", false),
                    config.getInt("wins", 0),
                    config.getInt("losses", 0),
                    file
            )
            ceraiaPlayers.add(ceraiaPlayer)
        }
    }

    fun getCeraiaPlayer(playerUUID: UUID): CeraiaPlayer {
        return ceraiaPlayers.find { it.uuid == playerUUID } ?: createNewCeraiaPlayer(playerUUID).also {
            ceraiaPlayers.add(it)
        }
    }

    fun getCeraiaPlayer(playerName: String): CeraiaPlayer {
        return ceraiaPlayers.find { it.name == playerName }
            ?: createNewCeraiaPlayer(Bukkit.getPlayer(playerName)?.uniqueId ?: throw IllegalArgumentException("Player not found"))
                .also { ceraiaPlayers.add(it) }
    }

    fun getCeraiaPlayer(player: Player): CeraiaPlayer {
        return ceraiaPlayers.find { it.uuid == player.uniqueId } ?: createNewCeraiaPlayer(player.uniqueId).also {
            ceraiaPlayers.add(it)
        }
    }

    private fun createNewCeraiaPlayer(playerUUID: UUID): CeraiaPlayer {
        val playerName = Bukkit.getPlayer(playerUUID)?.name ?: throw IllegalArgumentException("Player not found")
        val configFile = File(plugin.dataFolder, "data/users/$playerUUID.yml")
        try {
            configFile.createNewFile()
            val config = YamlConfiguration.loadConfiguration(configFile)

            val defaultElo = 1500

            config.apply {
                set("name", playerName)
                set("race", "human")
                set("faction", null)
                set("married", null)
                set("uuid", playerUUID.toString())
                set("elo", defaultElo)
                set("arenabanned", false)
                set("pvpbanned", false)
                set("wins", 0)
                set("losses", 0)
                set("logs", mutableListOf<String>())
            }
            config.save(configFile)

            return CeraiaPlayer(
                    plugin,
                    playerName,
                    "human",
                    null,
                    null,
                    playerUUID,
                    defaultElo,
                    false,
                    false,
                    0,
                    0,
                    configFile
            )
        } catch (e: IOException) {
            e.printStackTrace()
            return CeraiaPlayer(
                    plugin,
                    playerName,
                    "human",
                    null,
                    null,
                    playerUUID,
                    1500,
                    false,
                    false,
                    0,
                    0,
                    configFile
            )
        }
    }

    fun playerKill(playerKiller: UUID, playerVictim: UUID) {
        val killer = getCeraiaPlayer(playerKiller)
        val victim = getCeraiaPlayer(playerVictim)

        // Constants for the Elo calculation
        val kFactor = 32.0
        val expectedScoreKiller = 1.0 / (1.0 + Math.pow(10.0, (victim.elo - killer.elo) / 400.0))
        val expectedScoreVictim = 1.0 / (1.0 + Math.pow(10.0, (killer.elo - victim.elo) / 400.0))

        // Update Elo ratings
        val newEloKiller = (killer.elo + kFactor * (1.0 - expectedScoreKiller)).toInt()
        val newEloVictim = (victim.elo + kFactor * (0.0 - expectedScoreVictim)).toInt()

        // Set new Elo ratings
        killer.elo = newEloKiller
        victim.elo = newEloVictim

        // Update scoreboard
        val scoreboardManager: ScoreboardManager = Bukkit.getScoreboardManager()
        val scoreboardDefault: Scoreboard = scoreboardManager.newScoreboard
        val objectivePlayerList: Objective = scoreboardDefault.registerNewObjective("eloObjectivePlayerList", "dummy", MiniMessage.miniMessage().deserialize("Top Arena Players"))
        val objectiveBelowName: Objective = scoreboardDefault.registerNewObjective("eloObjectiveBelowName", "dummy", MiniMessage.miniMessage().deserialize("<green>ELO"))

        // Get all online players and set their score to their Elo rating
        Bukkit.getOnlinePlayers().forEach { onlinePlayer ->
            val doublePlayer = plugin.playerManager.getCeraiaPlayer(onlinePlayer.uniqueId)
            objectivePlayerList.getScore(onlinePlayer.name).score = doublePlayer.elo
            objectiveBelowName.getScore(onlinePlayer.name).score = doublePlayer.elo
            objectivePlayerList.displaySlot = DisplaySlot.PLAYER_LIST
            objectiveBelowName.displaySlot = DisplaySlot.BELOW_NAME
            onlinePlayer.scoreboard = scoreboardDefault
        }

        // Update player stats
        killer.savePlayer()
        victim.savePlayer()
    }

    fun calculateWinChance(playerKiller: UUID, playerVictim: UUID): Int {
        val killer = getCeraiaPlayer(playerKiller)
        val victim = getCeraiaPlayer(playerVictim)

        val expectedScoreKiller = 1.0 / (1.0 + Math.pow(10.0, (victim.elo - killer.elo) / 400.0))
        return (expectedScoreKiller * 100).toInt()
    }

    fun calculateLossChance(playerKiller: UUID, playerVictim: UUID): Int {
        val killer = getCeraiaPlayer(playerKiller)
        val victim = getCeraiaPlayer(playerVictim)

        val expectedScoreVictim = 1.0 / (1.0 + Math.pow(10.0, (killer.elo - victim.elo) / 400.0))
        return (expectedScoreVictim * 100).toInt()
    }

    fun getPlayer(uniqueId: UUID): CeraiaPlayer? {
        return ceraiaPlayers.find { it.uuid == uniqueId }
    }

    fun getPlayer(name: String): CeraiaPlayer? {
        return ceraiaPlayers.find { it.name.equals(name, ignoreCase = true) }
    }

    fun savePlayers() {
        plugin.logger.info("Saving players...")
        ceraiaPlayers.forEach { it.savePlayer() }
    }
}
